sw/device/mask_rom/boot.md

```diff
@@ -1,71 +1,134 @@
+
+    // Verify ROM_EXT Image Signature (2.c.iii)
+    // **Open Q:** Integration with Secure Boot Hardware
+    valid, signature = verify_rom_ext_signature(current_rom_ext_manifest);
+    if (!valid) {
+      // Boot Failure (check policy)
+      if (try_next_on_signature_failed(boot_policy))
+        continue
+      else
+        break
+    }
+
+    // System State Measurements (2.c.iv)
+    measurements = perform_system_state_measurements();
+    if (!boot_allowed(measurements)) {
+      // Lifecycle failure (no policy check)
+      break
+    }
+
+    // CreatorRootKey (2.c.iv)
+    root_key_identity = derive_creator_root_key(measurements);
+    load_root_key(root_key_identity);
+
+    // PMP Region for ROM_EXT (2.c.v)
+    // **Open Q:** Integration with Secure Boot Hardware
+    // **Open Q:** Do we need to lock down Boot ROM at final jump?
+    pmp_unlock_rom_ext();
+
+    // Lock down Peripherals based on descriptors in rom_ext
+    peripheral_lockdown(current_rom_ext_manifest);
+
+    // Transfer Execution to ROM_EXT (2.c.vi)
+    // **Open Q:** Integration with Secure Boot Hardware
+    if (!final_jump_to_rom_ext(current_rom_ext_manifest, signature)) {
+      // Boot Failure (check_policy)
+      if (try_next_on_jump_failed(boot_policy))
+        continue
+      else
+        break
+    }
   }
 
-  // Verify Signature (2.c.iii)
-  valid, signature = verify_rom_ext_signature(rom_ext_manifest);
-  if (!valid) {
-    // Boot Failure (check policy)
-  }
+  // Boot failed for all ROM_EXTs allowed by boot policy
+  boot_failed(boot_policy);
+}
+```
 
-  // System State Measurements (2.c.iv)
-  measurements = perform_system_state_measurements();
-  if (!boot_allowed(measurements)) {
-    // Lifecycle failure (no policy check)
-  }
+## Cleaning Device State
 
-  // CreatorRootKey (2.c.iv)
-  root_key_identity = derive_creator_root_key(measurements);
-  load_root_key(root_key_identity);
+Part of this process is done before we can execute any C code. In particular, we
+have to zero all registers and all of the main RAM before we setup the CRT
+(which allows us to execute any C code). This has to be done in assembly, as we
+cannot execute C without setting up the CRT.
 
-  // PMP Region for ROM_EXT (2.c.v)
-  //   Open Q: Do we need to lock down Boot ROM at final jump?
-  pmp_unlock_rom_ext();
+Setting up the CRT involves: loading the `.data` and `.bss` sections, and
+setting up the stack and probably `gp`.
 
-  // Lock down Peripherals based on descriptors in rom_ext
-  peripheral_lockdown(rom_ext_manifest);
+We want to do as little as possible in the hand-written assembly, because it is
+so hard to verify and to test. Additionally, we want to use the DIFs (which are
+written in C) because we have verified them, rather than duplicating their
+functionality in assembly, if at all possible. So this means we wait until we
+are executing C to do the final parts of chip reset, especially parts that may
+depend on reading or writing any device state.
```

Comment:
Just a thought: AFAIK, none of the DIFs actually need .bss or .data to exist, since they're designed to be stateless. Do you think we should only do a *partial* CRT initialization before calling into certain DIFs? How big is the chance that Clang screws us anyway and generates mutable globals?

Reply:
Why is a partial CRT init desirable? Removing everything from RAM protects us from attacks that leave useful data in RAM post-boot.

Reply:
I think this creates questions about "what part of the CRT does xyz need to be init" for certain tasks. It's going to make changes to DIFs taxing, because we need to make sure our mental models of what the compiler is doing matches what the compiler is actually doing...